crates/runtime/src/mmap.rs

```diff
@@ -326,6 +341,226 @@ impl Mmap {
         Ok(())
     }
 
+    /// Returns a page-aligned offset + length pair delimiting the memory pages which
+    /// are currently populated without generating any extraneous page faults.
+    #[cfg(target_os = "linux")]
+    fn populated_range(
+        &self,
+        accessible_offset: usize,
+        accessible: usize,
+    ) -> Result<(usize, usize)> {
+        // Docs: https://www.kernel.org/doc/Documentation/vm/pagemap.txt
+        use std::io::{Read, Seek};
+        const PAGE_SIZE: usize = 4096;
+
+        assert_eq!(rustix::process::page_size(), 4096);
+
+        assert_eq!(accessible_offset % 4096, 0);
+        assert_eq!(accessible % 4096, 0);
+
+        let mut page_last_index = 0;
+        let mut page_first_index = None;
+        unsafe {
+            let mut fp = std::fs::File::open("/proc/self/pagemap")
+                .context("failed to open /proc/self/pagemap")?;
```

Comment:
This doesn't work if `/proc` isn't mounted. In addition it doesn't check `/proc` is actually a mounted procfs. Rustix has code to check if the `/proc` is sane. It doesn't seem like it is exported though.

Reply:
Good point.

I've made a PR to `rustix` here to export it: https://github.com/bytecodealliance/rustix/pull/174